dsl/src/commonMain/kotlin/com/freeletics/flowredux/dsl/flow/WhileInState.kt

```diff
@@ -0,0 +1,46 @@
+package com.freeletics.flowredux.dsl.flow
+
+import com.freeletics.flowredux.GetState
+import com.freeletics.flowredux.dsl.internal.Action
+import kotlinx.coroutines.CancellationException
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.flow.consumeAsFlow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.launch
+
+@ExperimentalCoroutinesApi
+internal fun <S, A> Flow<Action<S, A>>.whileInState(
+    isInState: (S) -> Boolean,
+    getState: GetState<S>,
+    transform: (Flow<Action<S, A>>) -> Flow<Action<S, A>>
+) = channelFlow {
+    var currentChannel: Channel<Action<S, A>>? = null
```

Comment:
Do we need a mutex to ensure no 2 actions are processed in parallel or is there auch a guarantee already in place under the hood by channelFlow?

Reply:
Flow itself is sequential and under the assumption that collectors are not thread safe. So when we collect the upstream here we only get the next action when the previous one was handled which happens after we `send` it into `currentChannel`. So we sequentially put them into the channel and it also applies to creating/closing the channel. For the processing itself we are collecting the channel as a flow again, so it's als sequential and ordering of the action processing then depends on the transform. For onAction specifically it depends on the `flatMapPolicy`, CONCAT and LATEST are not parallel, while MERGE is.